src/notebooks/controllers/commands/pickDocumentKernelSourceCommand.ts

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+'use strict';
+import { inject, injectable } from 'inversify';
+import { NotebookDocument, window } from 'vscode';
+import { IExtensionSingleActivationService } from '../../../platform/activation/types';
+import { ICommandManager } from '../../../platform/common/application/types';
+import { Commands } from '../../../platform/common/constants';
+import { ContextKey } from '../../../platform/common/contextKey';
+import { IConfigurationService, IDisposableRegistry } from '../../../platform/common/types';
+import { noop } from '../../../platform/common/utils/misc';
+import { INotebookKernelSourceSelector } from '../types';
+
+// Command that we will place into the kernel picker to determine what the controller source is for this document
+@injectable()
+export class PickDocumentKernelSourceCommand implements IExtensionSingleActivationService {
+    private showPickDocumentKernelSourceContext: ContextKey;
+    constructor(
+        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry,
+        @inject(ICommandManager) private readonly commandManager: ICommandManager,
+        @inject(IConfigurationService) private readonly configService: IConfigurationService,
+        @inject(INotebookKernelSourceSelector) private readonly kernelSourceSelector: INotebookKernelSourceSelector
+    ) {
+        // Context keys to control when these commands are shown
+        this.showPickDocumentKernelSourceContext = new ContextKey(
+            'jupyter.pickDocumentKernelSourceContext',
+            this.commandManager
+        );
+    }
+
+    public async activate(): Promise<void> {
+        // Register for config changes
+        this.disposables.push(this.configService.getSettings().onDidChange(this.updateVisibility));
+
+        // Register our command to execute
+        this.disposables.push(
+            this.commandManager.registerCommand(Commands.PickDocumentKernelSource, this.pickDocumentKernelSource, this)
+        );
+
+        this.updateVisibility();
+    }
+
+    private async pickDocumentKernelSource(_notebook?: NotebookDocument) {
+        // We want to get the context here from the command, but that needs a fix on the core side: https://github.com/microsoft/vscode/issues/161445
+        if (window.activeNotebookEditor && this.configService.getSettings().kernelPickerType === 'Insiders') {
+            await this.kernelSourceSelector.selectKernelSource(window.activeNotebookEditor.notebook);
+        }
+    }
+
+    // Only show this command if we are in our Insiders picker type
+    private updateVisibility() {
+        if (this.configService.getSettings().kernelPickerType === 'Insiders') {
```

Comment:
It seems we are not picking up the context key change in the kernel picker so after changing the setting, a window reload is required for it to work. We might need a fix in the core.